### PR TITLE
[JUJU-1745] Add support for the secret-remove hook

### PR DESCRIPTION
**Pull request by wallyworld**
The secret-remove hook is implemented.
Plus support for deleting individual revisions rather than the whole secret.
Add a bug fix for secret deletion to clean up the owner label.

When a new secret revision is added, or a consumer tracks a new revision, we check to see if there are any revisions that are now unused. If so, we notify the unit agent and it can run the secret-remove hook. The hook can choose to delete that now obsolete revision if it wants. Once notified, it will not be notified again that the revision is obsolete.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Deploy a charm and add a secret with a label. I added a hook which logged the secret details in the secret-remove hook.
Grant access to another charm and get the secret content in that charm.
Add a new revision.
Do a secret get --update in the other charm.
The secret owner should see the secret remove hook called for revision 1.
Add a new revision. No hook.
Add a new revision - see a hook for the X-1 revision which is now unused.
Delete the obsolete revision.
From the CLI, use the secrets command to list revisions.
Delete the secret entirely.
Create a new one with the same label.

### CODE REVIEW

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 1]
```diff
@@ -412,6 +420,51 @@ func (s *SecretsManagerAPI) WatchSecretsChanges(args params.Entities) (params.St
...
+		if tag.Kind() == names.ApplicationTagKind {
+			token := s.leadershipChecker.LeadershipCheck(appName, s.authTag.Id())
+			if err := token.Check(0, nil); err != nil {
+				return "", nil, errors.Trace(err)
+			}
+		}
+
+		w := s.secretsConsumer.WatchObsoleteRevisions(arg.Tag)
+		if obsolete, ok := <-w.Changes(); ok {
+			changes := make([]string, len(obsolete))
```
Comment (hpidcock) [inline apiserver/facades/agent/secretsmanager/secrets.go:448]: You know I love this more
`changes := append([]string(nil), obsolete...)`

File: state/secrets.go [hunk 2]
```diff
@@ -905,6 +965,139 @@ func (st *State) WatchConsumedSecretsChanges(consumer string) StringsWatcher {
...
+			secretMetadataCollection, closer := st.db().GetCollection(secretMetadataC)
+			defer closer()
+
+			secretRevisionsCollection, closer := st.db().GetCollection(secretRevisionsC)
+			defer closer()
+
+			var doc secretRevisionDoc
+			err := secretRevisionsCollection.FindId(key).One(&doc)
+			if err != nil {
+				return false
```
Comment (hpidcock) [inline state/secrets.go:984]: should we handle errors better here?

File: state/secrets.go [hunk 3]
```diff
@@ -905,6 +965,139 @@ func (st *State) WatchConsumedSecretsChanges(consumer string) StringsWatcher {
...
+			}
+			parts := strings.Split(st.localID(doc.DocID), "/")
+
+			var mdDoc secretMetadataDoc
+			err = secretMetadataCollection.FindId(parts[0]).Select(bson.D{{"owner-tag", 1}}).One(&mdDoc)
+			if err != nil {
+				return false
```
Comment (hpidcock) [inline state/secrets.go:991]: here too

**Review by hpidcock — APPROVED** [review]

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 4]
```diff
@@ -412,6 +420,51 @@ func (s *SecretsManagerAPI) WatchSecretsChanges(args params.Entities) (params.St
 	return results, nil
 }
 
+// WatchObsoleteRevisions sets up a watcher to notify of changes to obsolete secret revisions for the specified secrets.
+func (s *SecretsManagerAPI) WatchObsoleteRevisions(args params.Entities) (params.StringsWatchResults, error) {
+	results := params.StringsWatchResults{
+		Results: make([]params.StringsWatchResult, len(args.Entities)),
+	}
+	one := func(arg params.Entity) (string, []string, error) {
```
Comment (tlm) [inline apiserver/facades/agent/secretsmanager/secrets.go:428]: Can we name this better? I would be keen to just make it it's own function.

File: apiserver/facades/agent/secretsmanager/secrets.go (hunk 1 above)
Comment (tlm) [reply apiserver/facades/agent/secretsmanager/secrets.go:448]: Do we even need the copy. Can't obsolete just be returned here. I can't see a reason why we need to shift it.

File: rpc/params/secrets.go [hunk 5]
```diff
@@ -74,15 +74,15 @@ type UpdateSecretArg struct {
...
 }
 
-// SecretURIArg holds the args for identifying a secret.
-type SecretURIArg struct {
-	// URI identifies the secret.
-	URI string `json:"uri"`
+// DeleteSecretArg holds the args for deleting a secret.
+type DeleteSecretArg struct {
+	URI       string `json:"uri"`
+	Revisions []int  `json:"revision,omitempty"`
```
Comment (tlm) [inline rpc/params/secrets.go:85]: This tag should be revisions ?

File: state/secrets.go [hunk 6]
```diff
@@ -905,6 +965,139 @@ func (st *State) WatchConsumedSecretsChanges(consumer string) StringsWatcher {
...
+	})
+}
+
+// markObsoleteRevisionOps returns ops for marking any revisions which are currently
+// not being tracked by any consumers as obsolete.
+func (st *State) markObsoleteRevisionOps(uri *secrets.URI, exceptForConsumer string, exceptForRev int) ([]txn.Op, error) {
+	var obsoleteOps []txn.Op
+	revs, latest, err := st.getOrphanedSecretRevisions(uri, exceptForConsumer, exceptForRev)
+	if err != nil {
+		return nil, errors.Trace(err)
```
Comment (tlm) [inline state/secrets.go:1013]: Should we annotate this a bit better?

File: worker/uniter/secrets/secrets.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+	rw UnitStateReadWriter,
+	logger Logger,
+) (SecretStateTracker, error) {
+	a := &Secrets{
+		client:   client,
+		unitTag:  tag,
+		logger:   logger,
+		stateOps: NewStateOps(rw),
+	}
+	if err := a.init(); err != nil {
```
Comment (tlm) [inline worker/uniter/secrets/secrets.go:54]: This could just be written  as.

```
err := a.init()
return a, err
```

**Review by tlm — APPROVED** [review]

File: state/secrets.go (hunk 2 above)
Comment (wallyworld) [reply state/secrets.go:984]: these methods do not have error handling - they were designed to only allow filtering on the id value itself.
without rewriting the watcher code, we just assume no event is fired on error

File: state/secrets.go (hunk 3 above)
Comment (wallyworld) [reply state/secrets.go:991]: yeah, it's how our watchers are designed

File: apiserver/facades/agent/secretsmanager/secrets.go (hunk 1 above)
Comment (wallyworld) [reply apiserver/facades/agent/secretsmanager/secrets.go:448]: probably don't need it

File: apiserver/facades/agent/secretsmanager/secrets.go (hunk 4 above)
Comment (wallyworld) [reply apiserver/facades/agent/secretsmanager/secrets.go:428]: this follows established patterns used approx 40 times elsewhere in out apiserver code :-)
would prefer to be consistent

File: rpc/params/secrets.go (hunk 5 above)
Comment (wallyworld) [reply rpc/params/secrets.go:85]: oops

File: state/secrets.go (hunk 6 above)
Comment (wallyworld) [reply state/secrets.go:1013]: can do

File: worker/uniter/secrets/secrets.go (hunk 7 above)
Comment (wallyworld) [reply worker/uniter/secrets/secrets.go:54]: nah, cause then we could not be returning the nil value on error, which is the idiomatic convention wherever possible
the "a" is misnamed, should be "s"

File: apiserver/facades/agent/secretsmanager/secrets.go (hunk 4 above)
Comment (tlm) [reply apiserver/facades/agent/secretsmanager/secrets.go:428]: ok (sad panda)

**Pull request merged**